### PR TITLE
Replacing bytes_accessed with access_reward_earned

### LINKED ISSUE #134: Record COST_PER_BYTE, MAKER_PAYMENT, BACKEND_PAYMENT in Delivery struct
Payments for deliveries are executed at multiple times. The reserve receives its share up front. Makers and datatrusts only receive their shares after the delivery has been executed and utilization has been reported by the datatrust. This can lead to inconsistencies in payment if a reparameterization happens in this time window. For example consider a reparam which alters `COST_PER_BYTE`. If `COST_PER_BYTE` increases, it is possible there will not be enough funds to pay the datatrust or makers. Similar considerations hold for `MAKER_PAYMENT` and `BACKEND_PAYMENT`

To fix this issue, we need to record the `COST_PER_BYTE`, `MAKER_PAYMENT, and `BACKEND_PAYMENT` that were valid when the delivery was requested and use them for all payments associated with the delivery. This can be done by adding extra `cost_per_byte`, `maker_payment`, and `backend_payment` fields to the `Delivery` struct:

https://github.com/computablelabs/goest/blob/master/contracts/Datatrust.vy#L9

And using these value instead of calling `self.parameterizer.getCostPerByte()`, `self.parameterizer.getBackendPayment()` and `self.parameterizer.getMakerPayment()` in https://github.com/computablelabs/goest/blob/master/contracts/Datatrust.vy#L294 and https://github.com/computablelabs/goest/blob/master/contracts/Listing.vy#L174

### PR DESCRIPTION
**Pull request by rbharath**
This PR replaces `Datatrust.bytes_accessed` with `Datatrust.access_reward_earned`. This replacement fixes #134.

### CODE REVIEW

File: tests/creatorblocktest/scenariofive/scenario_five_test.go [hunk 1]
```diff
@@ -344,10 +344,9 @@ func TestFullSimulation(t *testing.T) {
 			context.Blockchain.Commit()
 
 			// Check the number of bytes accessed is stored correctly
-			accessBalAfter, _ := deployed.DatatrustContract.GetBytesAccessed(nil, listingHash)
-			accessBalAdded := accessBalAfter.Sub(accessBalAfter, accessBal)
-			if accessBalAdded.Cmp(creditPerListing) != 0 {
-				t.Errorf("Expected %v to be %v", accessBalAdded, creditPerListing)
+			accessBalAfter, _ := deployed.DatatrustContract.GetAccessRewardEarned(nil, listingHash)
```
Comment (rbharath) [inline tests/creatorblocktest/scenariofive/scenario_five_test.go]: This is the only test that wasn't just a name swap. The reason was that the old test explicitly checked for the change in bytes. The new test just checks for an increment in access reward earned.

File: tests/datatrusttest/delivery_functionality_test.go [hunk 2]
```diff
@@ -55,7 +55,7 @@ func TestRequestDelivery(t *testing.T) {
 
 	// the actual request for a delivery...
 	_, delErr := deployed.DatatrustContract.RequestDelivery(test.GetTxOpts(
-		context.AuthUser3, nil, big.NewInt(test.ONE_GWEI*2), 250000), query, big.NewInt(1024*1024)) // ~1MB
+		context.AuthUser3, nil, big.NewInt(test.ONE_GWEI*2), 500000), query, big.NewInt(1024*1024)) // ~1MB
```
Comment (rbharath) [inline tests/datatrusttest/delivery_functionality_test.go]: I had to boost some gas in order to get `RequestDelivery` to go through. The reason is we're now writing to more storage.

File: tests/datatrusttest/delivery_reparam_pay_test.go [hunk 3]
```diff
@@ -268,13 +268,13 @@ func TestRequestDeliveryReparamPay(t *testing.T) {
 	beBalNow, _ := deployed.EtherTokenContract.BalanceOf(nil, beAddr)
 	dtBalNow, _ := deployed.EtherTokenContract.BalanceOf(nil, deployed.DatatrustAddress)
 
-	// newest bal will not have increased
-	if beBalNow.Cmp(beBal) != 0 {
+	// newest bal will have increased
```
Comment (rbharath) [inline tests/datatrusttest/delivery_reparam_pay_test.go]: This is the test added in #171 to test for the #134 failure. It fails without these fixes in place.

File: contracts/Datatrust.vy [hunk 4]
```diff
@@ -248,33 +265,45 @@ def listingAccessed(listing: bytes32, delivery: bytes32, amount: uint256):
...
   """
   assert msg.sender == self.listing_address
-  clear(self.bytes_accessed[hash]) # clear before paying
+  clear(self.access_reward_earned[hash]) # clear before paying
   self.ether_token.transfer(self.reserve_address, fee)
   # NOTE bytes accessed claimed event published by Listing contract
 
+@private
+@constant
+def _getDatatrustFee(delivery: bytes32) -> wei_value:
```
Comment (robrobbins) [inline contracts/Datatrust.vy]: same as the other _private. remove this method and put the lines in the calling method. not worth the overhead.

there's quite a few spots where i've had to fight this urge as well.

File: contracts/Listing.vy [hunk 5]
```diff
@@ -162,26 +162,24 @@ def resolveApplication(hash: bytes32):
...
   """
   @notice Allows a listing owner to claim the rewards of listing access. These support
   the market and will be noted at the listing.supply (MarketToken)
   @param hash The listing identifier
   """
   assert msg.sender == self.listings[hash].owner
-  # the algo for maker payment is (accessed*cost)/(100/maker_pct)
-  accessed: uint256 = self.datatrust.getBytesAccessed(hash)
-  maker_fee: wei_value = (self.parameterizer.getCostPerByte() * accessed * self.parameterizer.getMakerPayment()) / 100
+  maker_fee: wei_value = self.datatrust.getAccessRewardEarned(hash)
```
Comment (robrobbins) [inline contracts/Listing.vy]: i don't think we need to query the Datatrust for something it already knows and then tell it back

Comment (robrobbins) [reply contracts/Listing.vy]: so yeah, needed for minting...

File: tests/datatrusttest/delivery_functionality_test.go (hunk 2 above)
Comment (robrobbins) [reply tests/datatrusttest/delivery_functionality_test.go]: typical

File: tests/datatrusttest/delivery_reparam_pay_test.go (hunk 3 above)
Comment (robrobbins) [reply tests/datatrusttest/delivery_reparam_pay_test.go]: good.

there's prob an accurate way to square the decreasing bytes_purchased with the increasing rewards_earned based on the current params.

one thing i liked about bytes_purchased/bytes_accessed is they were 1:1 converse.

bytes_purchased/rewards_earned are as well (just with a little math in between them)

let's fit in a spec that highlights that, unless one of these already does well enough.

Comment (rbharath) [reply tests/datatrusttest/delivery_reparam_pay_test.go]: Added in a couple tests that do the math to check that the actual payment squares with the number of bytes accessed.

File: contracts/Datatrust.vy [hunk 6]
```diff
@@ -214,6 +218,9 @@ def requestDelivery(hash: bytes32, amount: uint256):
   self.bytes_purchased[msg.sender] += amount # all purchases by this user. deducted from via listing access
   self.deliveries[hash].owner = msg.sender
   self.deliveries[hash].bytes_requested = amount
+  self.deliveries[hash].cost_per_byte = self.parameterizer.getCostPerByte()
```
[the remark this answers is not in the document]
Comment (rbharath) [reply contracts/Datatrust.vy]: Good suggestion. Done

File: contracts/Datatrust.vy [hunk 7]
```diff
@@ -235,6 +242,16 @@ def getDelivery(hash: bytes32) -> (address, uint256, uint256):
   return (self.deliveries[hash].owner, self.deliveries[hash].bytes_requested,
     self.deliveries[hash].bytes_delivered)
 
+@private
+@constant
+def _getAccessRewardForAmount(delivery: bytes32, amount: uint256) -> wei_value:
```
[the remark this answers is not in the document]
Comment (rbharath) [reply contracts/Datatrust.vy]: Done

File: contracts/Datatrust.vy (hunk 4 above)
Comment (rbharath) [reply contracts/Datatrust.vy]: Done

File: tests/datatrusttest/delivery_functionality_test.go [hunk 8]
```diff
@@ -363,19 +374,35 @@ func TestClaimBytesAccessed(t *testing.T) {
 	// note the current datatrust banked eth token amount, at this point it should only be the maker split(s) from the outstanding bytes accessed
 	dataBal, _ := deployed.EtherTokenContract.BalanceOf(nil, deployed.DatatrustAddress)
 	// claim the listing for 1
-	_, clErr := deployed.ListingContract.ClaimBytesAccessed(test.GetTxOpts(context.AuthUser2, nil,
+	_, clErr := deployed.ListingContract.ClaimAccessReward(test.GetTxOpts(context.AuthUser2, nil,
 		big.NewInt(test.ONE_GWEI*2), 250000), listingHash1)
 	test.IfNotNil(t, clErr, "Error claiming access")
 	context.Blockchain.Commit()
 
+	// Compute Listing Reward per listing
```
Comment (rbharath) [inline tests/datatrusttest/delivery_functionality_test.go]: This checks that the access reward for datatrust corresponds to bytes accessed according to formula

File: tests/datatrusttest/delivery_reparam_pay_test.go [hunk 9]
```diff
@@ -268,13 +273,26 @@ func TestRequestDeliveryReparamPay(t *testing.T) {
...
+	if beBalNow.Cmp(beBal) != 1 {
 		t.Errorf("Expected %v to be %v", beBalNow, beBal)
 	}
 
-	// newest bal will not have decreased
-	if dtBalNow.Cmp(dtBal) != 0 {
+	// Compute theoretic backend payment
```
Comment (rbharath) [inline tests/datatrusttest/delivery_reparam_pay_test.go]: This checks that the access reward for datatrust corresponds to bytes accessed according to formula

File: tests/datatrusttest/delivery_functionality_test.go [hunk 10]
```diff
@@ -363,19 +374,35 @@ func TestClaimBytesAccessed(t *testing.T) {
...
+	costPerByte, _ := deployed.ParameterizerContract.GetCostPerByte(nil)
+	makerPayment, _ := deployed.ParameterizerContract.GetMakerPayment(nil)
+	amount := big.NewInt(1024 * 512)
+	// (costPerByte * amount * makerPayment) / 100
+	accessPayment := big.NewInt(1)
```
Comment (robrobbins) [inline tests/datatrusttest/delivery_functionality_test.go]: pretty sure you can chain these `mul().mul().div()` etc...

File: tests/datatrusttest/delivery_reparam_pay_test.go [hunk 11]
```diff
@@ -268,13 +273,26 @@ func TestRequestDeliveryReparamPay(t *testing.T) {
...
+	// (costPerByte * requested * backendPayment) / 100
+	bePayment := big.NewInt(1)
+	bePayment = bePayment.Mul(bePayment, costPerByte)
```
Comment (robrobbins) [inline tests/datatrusttest/delivery_reparam_pay_test.go]: same

File: contracts/Datatrust.vy [hunk 12]
```diff
@@ -248,34 +256,40 @@ def listingAccessed(listing: bytes32, delivery: bytes32, amount: uint256):
   assert msg.sender == self.backend_address
   assert self.data_hashes[listing] != EMPTY_BYTES32
   # this can be claimed later by the listing owner, and are subtractive to bytes_purchased
-  self.bytes_accessed[listing] += amount
   self.bytes_purchased[self.deliveries[delivery].owner] -= amount
   # bytes_delivered must eq (or exceed) bytes_requested in order for a datatrust to claim delivery
   self.deliveries[delivery].bytes_delivered += amount
+  # Compute the maker fee earned
+  cost_per_byte: wei_value = self.deliveries[delivery].cost_per_byte
+  maker_payment: uint256 = self.deliveries[delivery].maker_payment
```
Comment (robrobbins) [inline contracts/Datatrust.vy]: don't store vars if they are only ever used once. this is a pattern i have tried to enforce everywhere in the contracts. no reason to pay gas for the storage unless its used >= twice (double check any other new code for this too)

File: contracts/Datatrust.vy [hunk 13]
```diff
@@ -289,9 +303,14 @@ def delivered(delivery: bytes32, url: bytes32):
   owner: address = self.deliveries[delivery].owner
   requested: uint256 = self.deliveries[delivery].bytes_requested
   assert self.deliveries[delivery].bytes_delivered >= requested
-  # clear the delivery record first
+  cost_per_byte: wei_value = self.deliveries[delivery].cost_per_byte
+  backend_payment: uint256 = self.deliveries[delivery].backend_payment
```
Comment (robrobbins) [inline contracts/Datatrust.vy]: no reason to set these 2 as vars (same as above i think)

File: contracts/Datatrust.vy [hunk 14]
```diff
@@ -289,9 +303,14 @@ def delivered(delivery: bytes32, url: bytes32):
...
+  # The rounding issues are fine since at most 99 wei can be lost in
+  # rounding
+  back_fee: wei_value = (cost_per_byte * requested * backend_payment) / 100
+  #back_fee: wei_value = self._getDatatrustFee(delivery)
```
Comment (robrobbins) [inline contracts/Datatrust.vy]: remove this commented out line

File: tests/datatrusttest/delivery_functionality_test.go (hunk 10 above)
Comment (rbharath) [reply tests/datatrusttest/delivery_functionality_test.go]: Done

File: tests/datatrusttest/delivery_reparam_pay_test.go (hunk 11 above)
Comment (rbharath) [reply tests/datatrusttest/delivery_reparam_pay_test.go]: Done

File: contracts/Datatrust.vy (hunk 12 above)
Comment (rbharath) [reply contracts/Datatrust.vy]: Done. Double checked the PR to make sure no other double uses

File: contracts/Datatrust.vy (hunk 13 above)
Comment (rbharath) [reply contracts/Datatrust.vy]: Done

File: contracts/Datatrust.vy (hunk 14 above)
Comment (rbharath) [reply contracts/Datatrust.vy]: Done

**Pull request merged**